comtypes/npsupport.py

```diff
@@ -1,125 +1,158 @@
+
+    Returns None if the dtype cannot be created.
+
+    """
+    numpy = get_numpy()
+
+    # pointer typecode
+    ptr_typecode = '<u8' if is_64bits else '<u4'
+
+    _tagBRECORD_format = [
+        ('pvRecord', ptr_typecode),
+        ('pRecInfo', ptr_typecode),
+    ]
+
+    # overlapping typecodes only allowed in numpy version 1.7 or greater
+    U_VARIANT_format = dict(
+        names=[
+            'VT_BOOL', 'VT_I1', 'VT_I2', 'VT_I4', 'VT_I8', 'VT_INT', 'VT_UI1',
+            'VT_UI2', 'VT_UI4', 'VT_UI8', 'VT_UINT', 'VT_R4', 'VT_R8', 'VT_CY',
+            'c_wchar_p', 'c_void_p', 'pparray', 'bstrVal', '_tagBRECORD',
+        ],
+        formats=[
+            '<i2', '<i1', '<i2', '<i4', '<i8', '<i4', '<u1', '<u2', '<u4',
+            '<u8', '<u4', '<f4', '<f8', '<i8', ptr_typecode, ptr_typecode,
+            ptr_typecode, ptr_typecode, _tagBRECORD_format,
+        ],
+        offsets=[0] * 19  # This is what makes it a union
+    )
+
+    tagVARIANT_format = [
+        ("vt", '<u2'),
+        ("wReserved1", '<u2'),
+        ("wReserved2", '<u2'),
+        ("wReserved3", '<u2'),
+        ("_", U_VARIANT_format),
+    ]
+
+    return numpy.dtype(tagVARIANT_format)
+
+
+def isndarray(value):
+    """ Check if a value is an ndarray.
+
+    This cannot succeed if numpy is not available.
+
+    """
+    if not HAVE_NUMPY:
+        if hasattr(value, "__array_interface__"):
+            raise ValueError(
+                (
+                    "Argument {0} appears to be a numpy.ndarray, but "
+                    "comtypes numpy support has not been enabled. Please "
+                    "try calling comtypes.npsupport.enable_numpy_interop() "
+                    "before passing ndarrays as parameters."
+                ).format(value)
+            )
+        return False
+    numpy = get_numpy()
+    return isinstance(value, numpy.ndarray)
+
+
+def isdatetime64(value):
+    """ Check if a value is a datetime64.
+
+    This cannot succeed if datetime64 is not available.
+
+    """
+    if not HAVE_NUMPY:
+        return False
+    return isinstance(value, datetime64)
+
+
+def _check_ctypeslib_typecodes():
+    import numpy as np
+    from numpy import ctypeslib
+    try:
+        from numpy.ctypeslib import _typecodes
+    except ImportError:
+        from numpy.ctypeslib import as_ctypes_type
+
+        dtypes_to_ctypes = {}
+
+        for tp in set(np.sctypeDict.values()):
+            try:
+                ctype_for = as_ctypes_type(tp)
+                dtypes_to_ctypes[np.dtype(tp).str] = ctype_for
+            except NotImplementedError:
+                continue
+        ctypeslib._typecodes = dtypes_to_ctypes
+    return ctypeslib._typecodes
+
+
+def enable_numpy_interop():
+    """ Import the numpy library (if not already imported) and set up the
+    necessary functions for comtypes to work with ndarrays.
+
+    """
+    global HAVE_NUMPY
+    if HAVE_NUMPY:
+        return
+
+    global numpy
```

Comment:
Too many `global` keywords usually tells about weak design. Typical way to avoid it is a class with static members. The reason for `global` keyword here is seamless integration of `npsupport.numpy` module object for external user, but internal code uses `get_numpy()` function instead which is also public (not `_get_numpy()`).

Reply:
I agree that this global is unnecessary, I used that before realising that I needed `get_numpy()` so that I could raise the error if it wasn't enabled. However, there were still a lot of `global`s for the other objects in the top level namespace. I am not completely sure about whether a class approach is better, certainly we don't really want people instantiating such a class, but I have pushed an alternative design with an Interop class that has a single instance (called `interop`) created in the npsupport namespace, thus allowing e.g. npsupport.interop.enable() etc.
A potential improvement to this might be to rename npsupport.py to _npsupport.py and then have in __init__.py `from _npsupport import interop as npsupport` which would drop the extra level of indirection to get to `enable()`?

Reply:
It's a class with static members and static methods. No need to instantiate it. This is good idea to `import interop as npsupport` in `__init__.py`. Would `from comtypes import npsupport` work in this case? If yes, that's perfect!

Reply:
So the one reason to instantiate the class is for the `numpy` property - chaining @property and @classmethod is removed in Python 3.11 (see e.g. https://stackoverflow.com/a/64738850) so to allow this to be a property rather than e.g. a `get_numpy()` class method, we need the class to be instantiated. Is there a downside to instantiation in this case?
I will check the import situation but I believe that is how it would work.

Reply:
If there is a static member `interop.numpy` (not a property, just a member), it shouldn't require `@property` or `@classmethod` decorators. I would think about it so:
```python
class interop(object):
    numpy = None

    @staticmethod
    def enable():
        interop.numpy = __import__("numpy")

interop.enable()
print(interop.numpy)
# <module 'numpy' from 'C:\PythonXX\lib\site-packages\numpy\__init__.pyc'>
````

Reply:
I can, of course, do it this way - but the key point that I wanted to achieve with the property was to raise an ImportError with a helpful error message about how to enable numpy interop if it wasn't already enabled. If we decide that that message is not important, then I can do it as you suggest.

Reply:
OK, agree. I forgot this point.